File: Dynamic-Programming/646-M-Maximum-Length-of-Pair-Chain/max_len_of_pair_chain.py
class Solution:
    def findLongestChain(self, pairs: list[list[int]]) -> int:
        n = len(pairs)
        if n <= 1:
            return n
        
        pairs.sort(key=lambda x:x[1])

        OPT = [1] * n

        for i in range(n-2, -1, -1):
            for j in range(i+1, n):
                if pairs[i][1] < pairs[j][0]:
                    OPT[i] = max(OPT[i], OPT[j]+1)
        return max(OPT)

File: Dynamic-Programming/646-M-Maximum-Length-of-Pair-Chain/max_len_of_pair_chain.cpp
class Solution {
public:
    int findLongestChain(vector<vector<int>>& pairs) {
        int n = pairs.size();
        if (n <= 1) return n;

        sort(pairs.begin(), pairs.end());
        vector<int> OPT(n, 1);

        int result = 0;
        for (int i=n-2; i>=0; i--) {
            for (int j=1+i; j<n; j++) {
                if (pairs[i][1] < pairs[j][0]) {
                    OPT[i] = max(OPT[i], OPT[j] + 1);
                    if (result < OPT[i])
                        result = OPT[i];
                }
            }
        }
        return result;
    }
};